scipy/stats/_AB_SW.pyx

```diff
@@ -0,0 +1,463 @@
+    cdef double ZM = 0.17509e1, ZSS = 0.56268
+    cdef double BF1 = 0.8378, XX90 = 0.556, XX95 = 0.622
+    cdef double SQRTH = sqrt(2)/2, PI6 = 6/np.pi, PI_OVER_3 = np.pi/3
+    cdef double SMALL=1e-19, w, pw, an, an25, summ2, ssumm2, rsn
+    cdef double A1, A2, fac, delta, w1, y, ld, bf, gamma, m, s
+    cdef double RANGE, SA, SX, SSX, SSA, SAX, ASA, XSX, SSASSX, XX, XI
+    cdef double Z90F, Z95F, Z99F, ZFM, ZSD, ZBAR
+
+    if n1 < 0:
+        n1 = n
+    nn2 = n // 2
+    if nn2 < n2:
+        return 1., 1., 3
+    if n < 3:
+        return 1., 1., 1
+    w = 1.
+    pw = 1.
+    an = n
+
+
+    if not init:
+        if n == 3:
+            a[0] = SQRTH
+        else:
+            an25 = an + 0.25
+            summ2 = 0.
+            for ind1 in range(n2):
+                temp = _ppnd((ind1+1-0.375)/an25)
+                a[ind1] = temp
+                summ2 += temp**2
+
+            summ2 *= 2.
+            ssumm2 = sqrt(summ2)
+            rsn = 1 / sqrt(an)
+            A1 = _poly(c1, 6, rsn) - (a[0]/ssumm2)
+            if n > 5:
+                i1 = 2
+                A2 = -a[1]/ssumm2 + _poly(c2, 6, rsn)
+                fac = sqrt((summ2 - (2 * a[0]**2) - 2 * a[1]**2) / (1 - (2 * A1**2) - 2 * A2**2))
+                a[1] = A2
+            else:
+                i1 = 1
+                fac = sqrt((summ2 - 2 * a[0]**2)/(1 - 2 * A1**2))
+
+            a[0] = A1
+            for ind1 in range(i1, nn2):
+                a[ind1] *= -1./fac
+        init = True
+
+
+    if n1 < 3:
+        return w, pw, 1
+    ncens = n - n1
+
+    if ncens < 0 or ((ncens > 0) and (n < 20)):
+        return w, pw, 4
+    delta = ncens / an
+    if delta > 0.8:
+        return w, pw, 5
+
+
+    # Original Fortran code checked the W input value
+    # and acted on it via
+    # C
+    # C If W input as negative, calculate significance level of -W
+    # C
+    #       IF (W .LT. ZERO) THEN
+    #         W1 = ONE + W
+    #         IFAULT = 0
+    #         GOTO 70
+    #       END IF
+    # However W was marked as output in the fortran wrappers
+    # leading to undefined behavior of the uninitialized W
+    # Here W is assumed to be 0.
+    RANGE = x[n1-1] - x[0]
+    if RANGE < SMALL:
+        return w, pw, 6
+
+    XX = x[0] / RANGE
+    SX = XX
+    SA = -a[0]
+    ind2 = n - 2
+    for ind1 in range(1, n1):
+        XI = x[ind1] / RANGE
+        SX += XI
+        if ind1 != ind2:
+            SA += (-1 if ind1 < ind2 else 1)*a[min(ind1, ind2)]
+        XX = XI
+        ind2 -= 1
+
+    ifault = 0
+    if n > 5000:
+        ifault = 2
+
+    SA /= n1
+    SX /= n1
+    SSA, SSX, SAX = 0., 0., 0.
+    ind2 = n - 1
+    for ind1 in range(n1):
+        if ind1 != ind2:
+            ASA = (-1 if ind1 < ind2 else 1)*a[min(ind1, ind2)] - SA
+        else:
+            ASA = -SA
+
+        XSX = x[ind1]/RANGE - SX
+        SSA += ASA * ASA
+        SSX += XSX * XSX
+        SAX += ASA * XSX
+        ind2 -= 1
+
+    SSASSX = sqrt(SSA * SSX)
```

Comment:
I hate to be be "that guy" but if we keep the original FORTRAN style, this will still be hard to maintain in my opinion. Are all capitalized variables here matrices or was that convention just taken from the original?

I am not a statistical testing guy though, so cannot judge how complex this would be to implement in plain numpy or readable cython so.

Reply:
All coming from the fortran code. I just wanted to give whoever is reading this a chance to compare with the fortran file. If I knew a bit more about the algorithm, I'd give it a go for a numpy-ized version but this is just to get out of the fortran burden. 

I think a stats person would recognize the steps here better and can modify things in place. Because I really don't get what is going on here. But on the upside, other than variable naming, it's all arithmetics thanks to being written in the olden days, where there is nothing available.

Reply:
By the way, the `mvndst` code, that I am attempting with pythran, is even worse to decipher which includes a manual cholesky decomposition and pivoting. So fingers crossed :)

Reply:
@ilayn don't worry about mvndst. It is basically done in gh-17410; it just hasn't been made public.

Reply:
Ah you mean, we don't need to convert it ?

Reply:
Indeed that looks like the case. I think I can get rid of that permuted cholesky part with a regular lapack machinery if you want. Ah I can't be happier that I don't need to go over the remaining code to be honest heheh.

Reply:
>  I think I can get rid of that permuted cholesky part with a regular lapack machinery if you want.

Probably. Here are a few thoughts. 

Eliminating Fortran code, especially code without any bug reports, is not yet a high priority to me. (Personally, I was hoping that we would replace our code with `git submodule`s maintained by the modern Fortran folks. Part of the problem with maintaining them is the language, but I think there are resource limitations, too, so it would be good to offload some of that work since there are others interested in doing it. Anyway, that explains why I am not very enthusiastic about some of these Fortran-replacement PRs.)

That said, pythranizing `_mvt.py` is something I wanted to do anyway. It would close an issue filed by a user (gh-16142) and make `multivariate_t` much faster. So I would be willing to review a PR that pythranizes `_mvt.py` and replaces use of `mvndst`. I would suggest that it start with only essential changes to minimize the diff. The main change I anticipate is that `rng` inputs need to be replaced with a pre-generated random array of the appropriate size (unless Pythran has added the ability to accept NumPy RNGs as input). I'd ask that code to be pythranized be left in `_qmvnt.py` and any code that is not to be pythranized be added to a new file or `multivariate.py`. 

After that is done, replacing the permuted Cholesky with LAPACK would be a welcome follow-up PR.

Reply:
I used to believe the same thing; why change if things are working without any issues. But as you mentioned a few of them, these are 50 year old code that is practically unreadable. And it's based on a dying tech (F77 not implying anything about newer Fortran). Hence after reading all these fortran codes for the last month, I can safely say that argument about stability won't fly anymore. These code need to be get rid of. They are stopping us to even test ideas: copy paste some code, try it out, and so on. Even LAPACK folks are questioning the future of fortran let alone these individual code. 

I'm skipping the actual issues with F77 which we have beaten up to death elsewhere. But in particular with mnv I meant only the particular cholesky part that I know what it is doing. The remaining parts, I really don't have any capacity to integrate.

Reply:
> The remaining parts, I really don't have any capacity to integrate.

Right. If I seemed to ask for something else, let me know what it was so I can clarify.

Reply:
> Personally, I was hoping that we would replace our code with `git submodule`s maintained by the modern Fortran folks.

I have so far seen some stated enthusiasm, but very few (if any) contributions on SciPy itself or SciPy-equivalent code. Outside of working on LFortran itself being able to compiler SciPy of course. So aside from the question of what's most desirable - that depends on code complexity I'd say - I don't see this happening any time soon.

The stats/special routines are overall pretty simple compared to large components like sparse solver libraries, and I don't think they are candidates for separate Fortran packages. So I'd say a 1:1 translation of them to Python/Cython/Pythran is a clear win.

Reply:
> The stats/special routines are overall pretty simple compared to large components like sparse solver libraries, and I don't think they are candidates for separate Fortran packages. So I'd say a 1:1 translation of them to Python/Cython/Pythran is a clear win

Yes, I shouldn't have brought up the modern Fortran stuff, since I wouldn't expect them to take over this simple stuff.

I agree that it's a win if no bugs are introduced and we can find a reviewer for it.

Reply:
At least for this PR, all tests I've done are double precision identical. The reason I'm confident about that is because I also compiled the file externally and run the tests against each other. 

So if there are no `if` branches I've missed, I'd say this is pretty much identical.

Reply:
Ok. I can't review the code, but I can write some tests shortly.